MM-19628: Fix race in (*registeredPlugin).State access

agnivade opened the pull request:
#### Summary
We used a pointer to an integer in the State variable, and therefore
setting a pointer value directly caused a race condition.

Since the registeredPlugin is a private struct, changing it to a normal int
creates fewer pointers and lets us to fix the race condition by simply setting
the integer value and then doing a registeredPlugins.Store.

The alternative to keeping the pointer would be to create a copy of each
member of a registeredPlugin and do a store again, which I feel is a roundabout
way to achieve this.


#### Ticket Link
https://mattermost.atlassian.net/browse/MM-19628

lieut-data commented on plugin/environment.go:
```diff
@@ -114,13 +114,14 @@ func (env *Environment) GetPluginState(id string) int {
 	if rp, ok := env.registeredPlugins.Load(id); ok {
-		*rp.(*registeredPlugin).State = state
+		rp.(*registeredPlugin).State = state
+		env.registeredPlugins.Store(id, rp)
```
Since we're storing pointers to a `registeredPlugin` in the map, it seems like this `Store` is redundant -- aren't we already mutating on the line above? (Even with `State` as a value type?)

Is this a potential cause for another data race? Should we consider storing value types of registeredPointers as well?

agnivade replied:
I believe you are right @lieut-data. While this PR fixes the race for the `.State` access, the race due to the pointer to `*registeredPlugin` is still there. This was actually uncovered by @streamer45 here https://mattermost.atlassian.net/browse/MM-21019.

I think the right fix to this is to increase the lock scope from the `sync.Map` to use a general mutex and guard the access to the registeredPlugin also. So `registeredPlugins` becomes a simple map, and we stick in a mutex to the `Environment` struct for methods which mess with the state.

This would actually be a good change, because I believe `sync.Map` was not the perfect choice for this. From the [documentation](https://golang.org/pkg/sync/#Map), it mentions that `sync.Map` is optimized for cases when a key is only written once and read many times across several goroutines, or when there are disjoint sets of keys. Both of which is inapplicable here. So I think this is actually good for the codebase.

How about I send another PR for this. so that we keep this PR just for the State variable access ?

lieut-data replied:
> it mentions that sync.Map is optimized for cases when a key is only written once and read many times across several goroutines

I actually think `sync.Map` is perfect for this use case -- the vast majority of cases are reads, and we only write infrequently: on startup, or when a user adds/removes a plugin. Both of these pale in comparison to the read load.

> I think the right fix to this is to increase the lock scope from the sync.Map to use a general mutex and guard the access to the registeredPlugin also.

We intentionally switched to `sync.Map` to avoid re-entrancy issues with `sync.RWutex`: https://github.com/mattermost/mattermost-server/pull/9167.

I /think/ we can eliminate the race altogether if we just abandon the need to maintain a pointer to a `registeredPlugin` and just store values in the map instead. That, combined with your change to have `State` be a simple value, and it seems like we can keep `sync.Map` and avoid the data races. Thoughts?

agnivade replied:
> I actually think sync.Map is perfect for this use case -- the vast majority of cases are reads,
Ooh okay, that's nice :+1: 

> 
I /think/ we can eliminate the race altogether if we just abandon the need to maintain a pointer to a registeredPlugin and just store values in the map instead. 
Sure, we can go ahead with that.

lieut-data replied:
@agnivade, I'm not seeing the changes discussed above to abandon the pointer to a `registeredPlugin` and just store values in the map instead. From what I can see, there's still a race here in that we're directly mutating the `State` value of an object that might also be being read. 

Should we switch to just storing values in the `sync.Map` instead? It's possible I'm misunderstanding something, though!

agnivade commented:
/update-branch

agnivade commented:
Bump @lieut-data

lieut-data commented:
@agnivade, taking a look now.

Also, see https://developers.mattermost.com/contribute/getting-started/code-review/#if-you-are-a-core-committer-seeking-a-review for more details, but I'd recommend re-requesting a review once feedback is addressed to put the review back on the reviewer's radar (i.e. and show up in our awesome GitHub plugin's daily reminder).

agnivade commented:
@lieut-data - Sorry for the confusion. I mentioned:

> How about I send another PR for this. so that we keep this PR just for the State variable access ?

I meant that this PR does indeed fix _this_ race, which is for the State variable access. I wanted to send a separate PR addressing your comments.

However, if you want me to do it in this PR, that is fine too.

lieut-data commented:
@agnivade, ah, I missed that comment in your earlier message. Thanks for the clarification!

I'm fine with addressing the larger problem in a second PR, but I'd advocate that we remove the redundant `Store` here, since there's no reason to write anything back to the `sync.Map` in this case: I fear it masks that the write is still taking place on the shared object.

agnivade commented:
Makes sense. Will do. :+1:

agnivade commented:
/update-branch

jasonblais commented:
@DHaussermann quick reminder to help with this PR review when you have time. Thanks! :)

agnivade commented:
/update-branch

DHaussermann commented:
@streamer45 I wasn't able to test this on my local with the instructions provided in the Jira task. 
Can you help QA by this please?

streamer45 commented:
> @streamer45 I wasn't able to test this on my local with the instructions provided in the Jira task.
> Can you help QA by this please?

Yes, I can QA this.

streamer45 commented:
@DHaussermann 

Done testing this. Couldn't reproduce the data race with this PR applied. Good to merge for me.
